src/main/java/org/spongepowered/common/event/manager/ListenerClassVisitor.java

```diff
@@ -0,0 +1,451 @@
+import org.objectweb.asm.ClassVisitor;
+import org.objectweb.asm.Label;
+import org.objectweb.asm.MethodVisitor;
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.Type;
+import org.spongepowered.api.event.Listener;
+import org.spongepowered.api.event.Order;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.annotation.Annotation;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class ListenerClassVisitor extends ClassVisitor {
+
+    public static final String LISTENER_DESCRIPTOR = Type.getDescriptor(Listener.class);
+
+    final List<DiscoveredMethod> foundListenerMethods = new LinkedList<>();
+    final Class<?> declaringClass;
+
+    private ListenerClassVisitor(final Class<?> handle) {
+        super(Opcodes.ASM9);
+        this.declaringClass = handle;
+    }
+
+    static List<DiscoveredMethod> getEventListenerMethods(final Class<?> handle) throws
+        IOException,
+        NoSuchMethodException {
+        final @Nullable InputStream classStream = handle.getClassLoader().getResourceAsStream(
+            handle.getName().replace(".", "/") + ".class");
+        if (classStream == null) {
+            throw new IOException("Could not find class " + handle.getName());
+        }
+        final ClassReader reader = new ClassReader(classStream);
+        final ListenerClassVisitor classVisitor = new ListenerClassVisitor(handle);
+        reader.accept(classVisitor, 0);
+        return classVisitor.foundListenerMethods();
+    }
+
+    @Override
+    public MethodVisitor visitMethod(
+        final int access, final String name, final String descriptor, final String signature, final String[] exceptions
+    ) {
+        return new ListenerMethodVisitor(this, name, descriptor, access, signature);
+    }
+
+    public List<DiscoveredMethod> foundListenerMethods() {
+        return Collections.unmodifiableList(this.foundListenerMethods);
+    }
+
+    static class ListenerMethodVisitor extends MethodVisitor {
+        private final ListenerClassVisitor classVisitor;
+        @MonotonicNonNull private DiscoveredMethod discoveredMethod;
+
+        public ListenerMethodVisitor(
+            final ListenerClassVisitor listenerClassVisitor, final String name, final String descriptor,
+            final int access,
+            @Nullable final String signature
+        ) {
+            super(Opcodes.ASM9);
+            this.classVisitor = listenerClassVisitor;
+            final Type[] type = Type.getArgumentTypes(descriptor);
+            final ListenerParameter[] parameters = new ListenerParameter[type.length];
+            for (int i = 0; i < type.length; i++) {
+                parameters[i] = new ListenerParameter(this.discoveredMethod, type[i]);
+            }
+            // TODO - figure out a cleaner way to do this, this works for:
```

Comment:
maybe take a look at ASM's `SignatureReader`? it should be able to capture the applicable type, and Geantyref has methods to create a runtime reflection `Type` ourselves

Reply:
That's fair, I did read briefly about using SignatureReader, but didn't connect the dots on making it work (code written at 2am isn't always the sanest).

I did take advantage of Geantyref's TypeFactory in a few places already for creating `Type` instances as necessary.

Reply:
Rewrote with SignatureVisitor on the parameter